## PR #25 — Partial fix still broken (`intervals=` → `extractions=` + wrap in Extraction)

@claude thanks for the first kwarg rename, but the underlying shape is also wrong. Still 1 failing test:

```
tests/test_langextract_tool.py::TestIntegration::test_agent_usage_pattern
E   AnnotatedDocument.__init__() got an unexpected keyword argument 'intervals'
```

### Root cause

`lx.data.AnnotatedDocument` takes `extractions=[Extraction(...)]`, not `intervals=[CharInterval(...)]`. Each `Extraction` wraps a `CharInterval` plus a class/label.

The working reference is already in `praisonaiagents/tools/langextract_tools.py:94-120` (PR #1424). Please port that exact pattern.

### Fix for `praisonai_tools/tools/langextract_tool.py` (`_create_annotated_document`)

```diff
 def _create_annotated_document(text: str, extractions: List[str], document_id: str):
-    """Create langextract AnnotatedDocument with extractions as CharIntervals."""
+    """Create langextract AnnotatedDocument with extractions as Extraction objects."""
     lx = _get_langextract()
     if not lx:
         return None
-    
-    # Find all extraction positions in the text
-    intervals = []
-    for extraction in extractions:
-        start_pos = 0
-        while True:
-            pos = text.find(extraction, start_pos)
-            if pos == -1:
-                break
-            intervals.append(lx.data.CharInterval(
-                start_pos=pos,
-                end_pos=pos + len(extraction)
-            ))
-            start_pos = pos + 1
-    
-    # Create annotated document
-    return lx.data.AnnotatedDocument(
-        document_id=document_id,
-        text=text,
-        intervals=intervals
-    )
+
+    # Find all extraction positions and wrap as Extraction objects
+    extraction_objects = []
+    for i, extraction_text in enumerate(extractions or []):
+        if not extraction_text.strip():
+            continue
+        start_pos = 0
+        while True:
+            pos = text.lower().find(extraction_text.lower(), start_pos)
+            if pos == -1:
+                break
+            extraction_objects.append(lx.data.Extraction(
+                extraction_class=f"extraction_{i}",
+                extraction_text=extraction_text,
+                char_interval=lx.data.CharInterval(
+                    start_pos=pos,
+                    end_pos=pos + len(extraction_text),
+                ),
+                attributes={
+                    "index": i,
+                    "original_text": extraction_text,
+                    "tool": "langextract_extract",
+                },
+            ))
+            start_pos = pos + 1
+
+    return lx.data.AnnotatedDocument(
+        document_id=document_id,
+        text=text,
+        extractions=extraction_objects,
+    )
```

### Verify

```bash
cd /Users/praison/PraisonAI-Tools
pytest tests/test_langextract_tool.py -v
```
Expected: **16 passed, 0 failed**.

Local confirmation: running PR #1424's `langextract_extract` (which uses this exact pattern) produces a **7,585-byte HTML with 2 extractions** — proof the shape is right.

I'll re-merge the moment tests go green.

--- a/contract-001.html
+++ b/contract-001.html
@@ -1,0 +1,1 @@
+<html>mock visualization</html>

--- a/financial-analysis.html
+++ b/financial-analysis.html
@@ -1,0 +1,189 @@
+<style>
+.lx-highlight { position: relative; border-radius:3px; padding:1px 2px;}
+.lx-highlight .lx-tooltip {
+  visibility: hidden;
+  opacity: 0;
+  transition: opacity 0.2s ease-in-out;
+  background: #333;
+  color: #fff;
+  text-align: left;
+  border-radius: 4px;
+  padding: 6px 8px;
+  position: absolute;
+  z-index: 1000;
+  bottom: 125%;
+  left: 50%;
+  transform: translateX(-50%);
+  font-size: 12px;
+  max-width: 240px;
+  white-space: normal;
+  box-shadow: 0 2px 6px rgba(0,0,0,0.3);
+}
+.lx-highlight:hover .lx-tooltip { visibility: visible; opacity:1; }
+.lx-animated-wrapper { max-width: 100%; font-family: Arial, sans-serif; }
+.lx-controls {
+  background: #fafafa; border: 1px solid #90caf9; border-radius: 8px;
+  padding: 12px; margin-bottom: 16px;
+}
+.lx-button-row {
+  display: flex; justify-content: center; gap: 8px; margin-bottom: 12px;
+}
+.lx-control-btn {
+  background: #4285f4; color: white; border: none; border-radius: 4px;
+  padding: 8px 16px; cursor: pointer; font-size: 13px; font-weight: 500;
+  transition: background-color 0.2s;
+}
+.lx-control-btn:hover { background: #3367d6; }
+.lx-progress-container {
+  margin-bottom: 8px;
+}
+.lx-progress-slider {
+  width: 100%; margin: 0; appearance: none; height: 6px;
+  background: #ddd; border-radius: 3px; outline: none;
+}
+.lx-progress-slider::-webkit-slider-thumb {
+  appearance: none; width: 18px; height: 18px; background: #4285f4;
+  border-radius: 50%; cursor: pointer;
+}
+.lx-progress-slider::-moz-range-thumb {
+  width: 18px; height: 18px; background: #4285f4; border-radius: 50%;
+  cursor: pointer; border: none;
+}
+.lx-status-text {
+  text-align: center; font-size: 12px; color: #666; margin-top: 4px;
+}
+.lx-text-window {
+  font-family: monospace; white-space: pre-wrap; border: 1px solid #90caf9;
+  padding: 12px; max-height: 260px; overflow-y: auto; margin-bottom: 12px;
+  line-height: 1.6;
+}
+.lx-attributes-panel {
+  background: #fafafa; border: 1px solid #90caf9; border-radius: 6px;
+  padding: 8px 10px; margin-top: 8px; font-size: 13px;
+}
+.lx-current-highlight {
+  border-bottom: 4px solid #ff4444;
+  font-weight: bold;
+  animation: lx-pulse 1s ease-in-out;
+}
+@keyframes lx-pulse {
+  0% { text-decoration-color: #ff4444; }
+  50% { text-decoration-color: #ff0000; }
+  100% { text-decoration-color: #ff4444; }
+}
+.lx-legend {
+  font-size: 12px; margin-bottom: 8px;
+  padding-bottom: 8px; border-bottom: 1px solid #e0e0e0;
+}
+.lx-label {
+  display: inline-block;
+  padding: 2px 4px;
+  border-radius: 3px;
+  margin-right: 4px;
+  color: #000;
+}
+.lx-attr-key {
+  font-weight: 600;
+  color: #1565c0;
+  letter-spacing: 0.3px;
+}
+.lx-attr-value {
+  font-weight: 400;
+  opacity: 0.85;
+  letter-spacing: 0.2px;
+}
+
+/* Add optimizations with larger fonts and better readability for GIFs */
+.lx-gif-optimized .lx-text-window { font-size: 16px; line-height: 1.8; }
+.lx-gif-optimized .lx-attributes-panel { font-size: 15px; }
+.lx-gif-optimized .lx-current-highlight { text-decoration-thickness: 4px; }
+</style>
+<div class="lx-animated-wrapper lx-gif-optimized">
+  <div class="lx-attributes-panel">
+    <div class="lx-legend">Highlights Legend: <span class="lx-label" style="background-color:#D2E3FC;">extraction_0</span> <span class="lx-label" style="background-color:#C8E6C9;">extraction_1</span> <span class="lx-label" style="background-color:#FEF0C3;">extraction_2</span></div>
+    <div id="attributesContainer"></div>
+  </div>
+  <div class="lx-text-window" id="textWindow">
+    The <span class="lx-highlight lx-current-highlight" data-idx="0" style="background-color:#FEF0C3;">quarterly report</span> shows revenue of <span class="lx-highlight" data-idx="1" style="background-color:#D2E3FC;">$1.2M</span> and profit of <span class="lx-highlight" data-idx="2" style="background-color:#C8E6C9;">$300K</span>.
+  </div>
+  <div class="lx-controls">
+    <div class="lx-button-row">
+      <button class="lx-control-btn" onclick="playPause()">▶️ Play</button>
+      <button class="lx-control-btn" onclick="prevExtraction()">⏮ Previous</button>
+      <button class="lx-control-btn" onclick="nextExtraction()">⏭ Next</button>
+    </div>
+    <div class="lx-progress-container">
+      <input type="range" id="progressSlider" class="lx-progress-slider"
+             min="0" max="2" value="0"
+             onchange="jumpToExtraction(this.value)">
+    </div>
+    <div class="lx-status-text">
+      Entity <span id="entityInfo">1/3</span> |
+      Pos <span id="posInfo">[38-43]</span>
+    </div>
+  </div>
+</div>
+
+<script>
+  (function() {
+    const extractions = [{"index": 0, "class": "extraction_2", "text": "quarterly report", "color": "#FEF0C3", "startPos": 4, "endPos": 20, "beforeText": "The ", "extractionText": "quarterly report", "afterText": " shows revenue of $1.2M and profit of $300K.", "attributesHtml": "<div><strong>class:</strong> extraction_2</div><div><strong>attributes:</strong> {<span class=\"lx-attr-key\">index</span>: <span class=\"lx-attr-value\">2</span>, <span class=\"lx-attr-key\">original_text</span>: <span class=\"lx-attr-value\">quarterly report</span>, <span class=\"lx-attr-key\">tool</span>: <span class=\"lx-attr-value\">langextract_extract</span>}</div>"}, {"index": 1, "class": "extraction_0", "text": "$1.2M", "color": "#D2E3FC", "startPos": 38, "endPos": 43, "beforeText": "The quarterly report shows revenue of ", "extractionText": "$1.2M", "afterText": " and profit of $300K.", "attributesHtml": "<div><strong>class:</strong> extraction_0</div><div><strong>attributes:</strong> {<span class=\"lx-attr-key\">index</span>: <span class=\"lx-attr-value\">0</span>, <span class=\"lx-attr-key\">original_text</span>: <span class=\"lx-attr-value\">$1.2M</span>, <span class=\"lx-attr-key\">tool</span>: <span class=\"lx-attr-value\">langextract_extract</span>}</div>"}, {"index": 2, "class": "extraction_1", "text": "$300K", "color": "#C8E6C9", "startPos": 58, "endPos": 63, "beforeText": "The quarterly report shows revenue of $1.2M and profit of ", "extractionText": "$300K", "afterText": ".", "attributesHtml": "<div><strong>class:</strong> extraction_1</div><div><strong>attributes:</strong> {<span class=\"lx-attr-key\">index</span>: <span class=\"lx-attr-value\">1</span>, <span class=\"lx-attr-key\">original_text</span>: <span class=\"lx-attr-value\">$300K</span>, <span class=\"lx-attr-key\">tool</span>: <span class=\"lx-attr-value\">langextract_extract</span>}</div>"}];
+    let currentIndex = 0;
+    let isPlaying = false;
+    let animationInterval = null;
+    let animationSpeed = 1.0;
+
+    function updateDisplay() {
+      const extraction = extractions[currentIndex];
+      if (!extraction) return;
+
+      document.getElementById('attributesContainer').innerHTML = extraction.attributesHtml;
+      document.getElementById('entityInfo').textContent = (currentIndex + 1) + '/' + extractions.length;
+      document.getElementById('posInfo').textContent = '[' + extraction.startPos + '-' + extraction.endPos + ']';
+      document.getElementById('progressSlider').value = currentIndex;
+
+      const playBtn = document.querySelector('.lx-control-btn');
+      if (playBtn) playBtn.textContent = isPlaying ? '⏸ Pause' : '▶️ Play';
+
+      const prevHighlight = document.querySelector('.lx-text-window .lx-current-highlight');
+      if (prevHighlight) prevHighlight.classList.remove('lx-current-highlight');
+      const currentSpan = document.querySelector('.lx-text-window span[data-idx="' + currentIndex + '"]');
+      if (currentSpan) {
+        currentSpan.classList.add('lx-current-highlight');
+        currentSpan.scrollIntoView({block: 'center', behavior: 'smooth'});
+      }
+    }
+
+    function nextExtraction() {
+      currentIndex = (currentIndex + 1) % extractions.length;
+      updateDisplay();
+    }
+
+    function prevExtraction() {
+      currentIndex = (currentIndex - 1 + extractions.length) % extractions.length;
+      updateDisplay();
+    }
+
+    function jumpToExtraction(index) {
+      currentIndex = parseInt(index);
+      updateDisplay();
+    }
+
+    function playPause() {
+      if (isPlaying) {
+        clearInterval(animationInterval);
+        isPlaying = false;
+      } else {
+        animationInterval = setInterval(nextExtraction, animationSpeed * 1000);
+        isPlaying = true;
+      }
+      updateDisplay();
+    }
+
+    window.playPause = playPause;
+    window.nextExtraction = nextExtraction;
+    window.prevExtraction = prevExtraction;
+    window.jumpToExtraction = jumpToExtraction;
+
+    updateDisplay();
+  })();
+</script>

--- a/praisonai_tools/tools/langextract_tool.py
+++ b/praisonai_tools/tools/langextract_tool.py
@@ -46,30 +46,40 @@ def _get_langextract():
 
 
 def _create_annotated_document(text: str, extractions: List[str], document_id: str):
-    """Create langextract AnnotatedDocument with extractions as CharIntervals."""
+    """Create langextract AnnotatedDocument with extractions as Extraction objects."""
     lx = _get_langextract()
     if not lx:
         return None
-    
-    # Find all extraction positions in the text
-    intervals = []
-    for extraction in extractions:
+
+    # Find all extraction positions and wrap as Extraction objects
+    extraction_objects = []
+    for i, extraction_text in enumerate(extractions or []):
+        if not extraction_text.strip():
+            continue
         start_pos = 0
         while True:
-            pos = text.find(extraction, start_pos)
+            pos = text.lower().find(extraction_text.lower(), start_pos)
             if pos == -1:
                 break
-            intervals.append(lx.data.CharInterval(
-                start_pos=pos,
-                end_pos=pos + len(extraction)
+            extraction_objects.append(lx.data.Extraction(
+                extraction_class=f"extraction_{i}",
+                extraction_text=extraction_text,
+                char_interval=lx.data.CharInterval(
+                    start_pos=pos,
+                    end_pos=pos + len(extraction_text),
+                ),
+                attributes={
+                    "index": i,
+                    "original_text": extraction_text,
+                    "tool": "langextract_extract",
+                },
             ))
             start_pos = pos + 1
-    
-    # Create annotated document
+
     return lx.data.AnnotatedDocument(
         document_id=document_id,
         text=text,
-        intervals=intervals
+        extractions=extraction_objects,
     )
 
 

--- a/praisonai_tools/tools/langextract_tool.py
+++ b/praisonai_tools/tools/langextract_tool.py
@@ -151,9 +151,26 @@ class LangExtractTool(BaseTool):
             
             output_path = Path(output_path).resolve()
             
-            # Save annotated documents and visualize
-            lx.io.save_annotated_documents([doc], str(output_path.parent / f"{document_id}.jsonl"))
-            lx.visualize(str(output_path))
+            # Save JSONL and generate HTML from it
+            import tempfile, os
+            jsonl_dir = tempfile.gettempdir()
+            jsonl_path = os.path.join(jsonl_dir, f"{document_id}.jsonl")
+            lx.io.save_annotated_documents(
+                [doc],
+                output_name=os.path.basename(jsonl_path),
+                output_dir=jsonl_dir,
+            )
+
+            html = lx.visualize(jsonl_path)
+            html_content = html.data if hasattr(html, "data") else html
+
+            output_path = Path(output_path).resolve() if output_path else Path(f"{document_id}.html").resolve()
+            output_path.write_text(html_content, encoding="utf-8")
+
+            try:
+                os.remove(jsonl_path)
+            except OSError:
+                pass
             
             result = {
                 "success": True,

--- a/tests/test_langextract_tool.py
+++ b/tests/test_langextract_tool.py
@@ -60,7 +60,7 @@ class TestLangExtractTool:
         
         # Mock visualization methods
         mock_lx.io.save_annotated_documents = Mock()
-        mock_lx.visualize = Mock()
+        mock_lx.visualize = Mock(return_value="<html>mock visualization</html>")
         
         text = "John Doe works at OpenAI"
         extractions = ["John Doe", "OpenAI"]
@@ -96,7 +96,7 @@ class TestLangExtractTool:
         with patch('praisonai_tools.tools.langextract_tool._create_annotated_document') as mock_create_doc:
             mock_create_doc.return_value = Mock()
             mock_lx.io.save_annotated_documents = Mock()
-            mock_lx.visualize = Mock()
+            mock_lx.visualize = Mock(return_value="<html>mock visualization</html>")
             
             with patch('webbrowser.open') as mock_open:
                 with tempfile.TemporaryDirectory() as temp_dir:
@@ -138,7 +138,7 @@ class TestLangExtractTool:
         
         # Mock visualization methods
         mock_lx.io.save_annotated_documents = Mock()
-        mock_lx.visualize = Mock()
+        mock_lx.visualize = Mock(return_value="<html>mock visualization</html>")
         
         # Create temporary file
         with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.txt') as temp_file:
@@ -256,7 +256,7 @@ class TestIntegration:
         mock_get_lx.return_value = mock_lx
         mock_create_doc.return_value = Mock()
         mock_lx.io.save_annotated_documents = Mock()
-        mock_lx.visualize = Mock()
+        mock_lx.visualize = Mock(return_value="<html>mock visualization</html>")
         
         # Simulate contract text with key terms
         contract_text = """
@@ -287,13 +287,13 @@ class TestIntegration:
         analysis_text = "The quarterly report shows revenue of $1.2M and profit of $300K."
         findings = ["$1.2M", "$300K", "quarterly report"]
         
-        # Without langextract installed, should get helpful error
         result = langextract_extract(
             text=analysis_text,
             extractions=findings,
-            document_id="financial-analysis"
+            document_id="financial-analysis",
+            auto_open=False,
         )
-        
-        # Should get error about missing langextract
-        assert "error" in result
-        assert "langextract not installed" in result["error"]
+
+        assert result["success"] is True
+        assert result["document_id"] == "financial-analysis"
+        assert result["extractions_count"] == len(findings)

--- a/tmpc8_hvm6m.html
+++ b/tmpc8_hvm6m.html
@@ -1,0 +1,1 @@
+<html>mock visualization</html>

